File: MyGameFly/MyGameFly/MainMenu.cpp
#include "MainMenu.h"
#include <iostream>

MainMenu::MainMenu(sf::Vector2u winSize)
    : selectedMode(GameMode::NONE), isActive(true), windowSize(winSize)
{
    // Load font - try the same paths as in main.cpp
    bool fontLoaded = false;

#ifdef _WIN32
    if (font.openFromFile("arial.ttf") ||
        font.openFromFile("C:/Windows/Fonts/arial.ttf") ||
        font.openFromFile("C:/Windows/Fonts/Arial.ttf")) {
        fontLoaded = true;
    }
#elif defined(__APPLE__)
    if (font.openFromFile("arial.ttf") ||
        font.openFromFile("/Library/Fonts/Arial.ttf") ||
        font.openFromFile("/System/Library/Fonts/Arial.ttf")) {
        fontLoaded = true;
    }
#elif defined(__linux__)
    if (font.openFromFile("arial.ttf") ||
        font.openFromFile("/usr/share/fonts/truetype/msttcorefonts/Arial.ttf") ||
        font.openFromFile("/usr/share/fonts/TTF/arial.ttf")) {
        fontLoaded = true;
    }
#else
    fontLoaded = font.openFromFile("arial.ttf");
#endif

    if (!fontLoaded) {
        std::cerr << "Warning: Could not load font file for menu. Text won't display correctly." << std::endl;
        // Create a dummy font or handle the error appropriately
        // For now, we'll continue but buttons won't have text
    }

    setupBackground();
    createButtons();
}

void MainMenu::setupBackground() {
    background.setSize(sf::Vector2f(static_cast<float>(windowSize.x), static_cast<float>(windowSize.y)));
    background.setFillColor(sf::Color(20, 20, 40, 200)); // Dark blue semi-transparent
    background.setPosition(sf::Vector2f(0.0f, 0.0f));
}

void MainMenu::createButtons() {
    float centerX = static_cast<float>(windowSize.x) / 2.0f;
    float startY = static_cast<float>(windowSize.y) / 2.0f - 100.0f; // Start a bit above center
    float buttonWidth = 200.0f;
    float buttonHeight = 50.0f;
    float spacing = 70.0f;

    // Only create buttons if we have a valid font
    try {
        // Single Player Button
        auto singlePlayerButton = std::make_unique<Button>(
            sf::Vector2f(centerX - buttonWidth / 2, startY),
            sf::Vector2f(buttonWidth, buttonHeight),
            "Single Player",
            font,
            [this]() { onSinglePlayerClicked(); }
        );
        buttons.push_back(std::move(singlePlayerButton));

        // Multiplayer Button
        auto multiplayerButton = std::make_unique<Button>(
            sf::Vector2f(centerX - buttonWidth / 2, startY + spacing),
            sf::Vector2f(buttonWidth, buttonHeight),
            "Multiplayer",
            font,
            [this]() { onMultiplayerClicked(); }
        );
        buttons.push_back(std::move(multiplayerButton));

        // Quit Button
        auto quitButton = std::make_unique<Button>(
            sf::Vector2f(centerX - buttonWidth / 2, startY + spacing * 2),
            sf::Vector2f(buttonWidth, buttonHeight),
            "Quit",
            font,
            [this]() { onQuitClicked(); }
        );
        buttons.push_back(std::move(quitButton));
    }
    catch (const std::exception& e) {
        std::cerr << "Error creating buttons: " << e.what() << std::endl;
        // If button creation fails, we could create simple colored rectangles instead
    }
}

void MainMenu::onSinglePlayerClicked() {
    selectedMode = GameMode::SINGLE_PLAYER;
    isActive = false;
    std::cout << "Single Player mode selected" << std::endl;
}

void MainMenu::onMultiplayerClicked() {
    selectedMode = GameMode::MULTIPLAYER_HOST; // This will trigger the multiplayer submenu
    isActive = false;
    std::cout << "Opening multiplayer menu..." << std::endl;
}

void MainMenu::onQuitClicked() {
    selectedMode = GameMode::QUIT;
    isActive = false;
    std::cout << "Quit selected" << std::endl;
}

void MainMenu::handleEvent(const sf::Event& event, const sf::Vector2f& mousePos) {
    if (!isActive) return;

    if (event.is<sf::Event::MouseButtonPressed>()) {
        const auto* mouseEvent = event.getIf<sf::Event::MouseButtonPressed>();
        if (mouseEvent && mouseEvent->button == sf::Mouse::Button::Left) {
            // Check which button was clicked
            for (auto& button : buttons) {
                if (button->contains(mousePos)) {
                    button->handleClick();
                    break;
                }
            }
        }
    }
}

void MainMenu::update(const sf::Vector2f& mousePos) {
    if (!isActive) return;

    // Update all buttons (for hover effects)
    for (auto& button : buttons) {
        button->update(mousePos);
    }
}

void MainMenu::draw(sf::RenderWindow& window) {
    if (!isActive) return;

    // Draw background
    window.draw(background);

    // Draw all buttons
    for (auto& button : buttons) {
        button->draw(window);
    }
}

File: KatieFlySimRust/src/menus/main_menu.rs
// Main Menu - Entry point menu
// Ported from C++ MainMenu class

use sfml::graphics::{Color, Font, RenderTarget, RenderWindow, Text, Transformable};
use sfml::system::Vector2f;
use sfml::window::mouse;

use crate::game_state::GameMode;
use crate::ui::Button;

/// Main menu with game mode selection
pub struct MainMenu<'a> {
    title: Text<'a>,
    single_player_button: Button<'a>,
    multiplayer_button: Button<'a>,
    quit_button: Button<'a>,
    selected_mode: GameMode,
}

impl<'a> MainMenu<'a> {
    pub fn new(window_size: Vector2f, font: &'a Font) -> Self {
        // Title
        let mut title = Text::new("KatieFlySimRust", font, 72);
        title.set_fill_color(Color::WHITE);
        let title_bounds = title.local_bounds();
        title.set_origin(Vector2f::new(title_bounds.width / 2.0, 0.0));
        title.set_position(Vector2f::new(window_size.x / 2.0, 100.0));

        // Button positioning
        let button_width = 300.0;
        let button_height = 60.0;
        let button_spacing = 80.0;
        let start_y = window_size.y / 2.0 - 50.0;

        // Single Player button
        let single_player_button = Button::new(
            Vector2f::new(
                window_size.x / 2.0 - button_width / 2.0,
                start_y,
            ),
            Vector2f::new(button_width, button_height),
            "Single Player",
            font,
            Color::rgb(50, 100, 150),
        );

        // Multiplayer button
        let multiplayer_button = Button::new(
            Vector2f::new(
                window_size.x / 2.0 - button_width / 2.0,
                start_y + button_spacing,
            ),
            Vector2f::new(button_width, button_height),
            "Multiplayer",
            font,
            Color::rgb(50, 120, 100),
        );

        // Quit button
        let quit_button = Button::new(
            Vector2f::new(
                window_size.x / 2.0 - button_width / 2.0,
                start_y + button_spacing * 2.0,
            ),
            Vector2f::new(button_width, button_height),
            "Quit",
            font,
            Color::rgb(120, 50, 50),
        );

        MainMenu {
            title,
            single_player_button,
            multiplayer_button,
            quit_button,
            selected_mode: GameMode::None,
        }
    }

    /// Update menu and handle input
    pub fn update(&mut self, window: &RenderWindow) -> GameMode {
        let mouse_pressed = mouse::Button::Left.is_pressed();

        // Update buttons
        if self.single_player_button.update(window, mouse_pressed) {
            self.selected_mode = GameMode::SinglePlayer;
            return GameMode::SinglePlayer;
        }

        if self.multiplayer_button.update(window, mouse_pressed) {
            self.selected_mode = GameMode::Multiplayer;
            return GameMode::Multiplayer;
        }

        if self.quit_button.update(window, mouse_pressed) {
            self.selected_mode = GameMode::Quit;
            return GameMode::Quit;
        }

        GameMode::None
    }

    /// Draw the menu
    pub fn draw(&self, window: &mut RenderWindow) {
        window.draw(&self.title);
        self.single_player_button.draw(window);
        self.multiplayer_button.draw(window);
        self.quit_button.draw(window);
    }

    /// Get the selected mode
    pub fn selected_mode(&self) -> GameMode {
        self.selected_mode
    }

    /// Reset selection
    pub fn reset(&mut self) {
        self.selected_mode = GameMode::None;
    }
}
